### ISSUE TITLE
Support for rich Enumerations

### ISSUE THREAD

**Issue by samreid** [phetsims/phet-core#50]
From https://github.com/phetsims/phet-core/issues/49 which came from https://github.com/phetsims/phet-core/issues/42

I said: 

> Or each Enumeration could be a type, then create values as instances of the type, then instanceof should work...

@jonathanolson replied:

> That sounds potentially nice. You'd want to support subclassing the class, so it might be a bit tricky to set up though.

@pixelzoom also said:

> I would never expect  to be able to use `Animals.COW instanceof Animals`, since Animals is not a class and Animals.COW is not an instance.  I would expect `Animals.includes( Animals.COW )`.

and @pixelzoom also said:

> `instanceof` works with enumerations in Java because `enum` is part of the language.  That's not the case here, we're rolling our own.  Are you suggesting that we should somehow overload `instanceof`?

Other Property validation strategies are discussed in #49, this issue is all about if/how we should support instanceof.

**Comment by samreid** [comment]
@jonathanolson can you please elaborate on:

> You'd want to support subclassing the class, so it might be a bit tricky to set up though.

Are you saying if I had `CardinalDirection` as an `Enumeration` "type" that it would be important to be able to subtype it?  Can you subtype Enums in other languages?

**Comment by jonathanolson** [comment]
> Are you saying if I had CardinalDirection as an Enumeration "type" that it would be important to be able to subtype it? Can you subtype Enums in other languages?

I was thinking in terms of rich enumerations, e.g. Orientation.

**Comment by jonathanolson** [comment]
Java allows rich enumerations, e.g.:

```js
public enum Planet {
    MERCURY (3.303e+23, 2.4397e6),
    VENUS   (4.869e+24, 6.0518e6),
    EARTH   (5.976e+24, 6.37814e6),
    MARS    (6.421e+23, 3.3972e6),
    JUPITER (1.9e+27,   7.1492e7),
    SATURN  (5.688e+26, 6.0268e7),
    URANUS  (8.686e+25, 2.5559e7),
    NEPTUNE (1.024e+26, 2.4746e7);

    private final double mass;   // in kilograms
    private final double radius; // in meters
    Planet(double mass, double radius) {
        this.mass = mass;
        this.radius = radius;
```

which is the equivalent of what I'd want.

**Comment by samreid** [comment]
Would this be sufficient?

```js
// Copyright 2018, University of Colorado Boulder

define( require => {
  'use strict';

  // modules
  const mySimulation = require( 'MY_SIMULATION/mySimulation' );

  let frozen = false;

  class Planet {

    /**
     * @param {number} mass
     * @param {number} radius
     */
    constructor( mass, radius ) {
      assert && assert( !frozen, 'Cannot create new instances of Planet' );
      this.mass = mass;
      this.radius = radius;
    }
  }

  Planet.VALUES = [];
  const append = planet => Planet.VALUES.push( planet );
  Planet.MERCURY = append( new Planet( 3.303e+23, 2.4397e6 ) );
  Planet.VENUS = append( new Planet( 4.869e+24, 6.0518e6 ) );
  Planet.EARTH = append( new Planet( 5.976e+24, 6.37814e6 ) );
  Planet.MARS = append( new Planet( 6.421e+23, 3.3972e6 ) );
  Planet.JUPITER = append( new Planet( 1.9e+27, 7.1492e7 ) );
  Planet.SATURN = append( new Planet( 5.688e+26, 6.0268e7 ) );
  Planet.URANUS = append( new Planet( 8.686e+25, 2.5559e7 ) );
  Planet.NEPTUNE = append( new Planet( 1.024e+26, 2.4746e7 ) );

  assert && Object.freeze( Planet );
  assert && Object.freeze( Planet.VALUES );
  frozen = true;

  return mySimulation.register( 'Planet', Planet );
} );
```

Would we benefit from library support for a pattern like that?

**Comment by jonathanolson** [comment]
> Would this be sufficient?
> Would we benefit from library support for a pattern like that?

The above type of thing would work, but it seems like it's duplicating what Enumeration does, and missing things like Enumeration's includes() or whatever we add to Enumeration in the future. Ideally Enumeration itself should work with the rich `Planet`.

**Comment by samreid** [comment]
In https://github.com/phetsims/phet-core/blob/0406bbb48eae39fccfcb1ba49b29767496ac3ce0/js/Orientation.js @jonathanolson and I committed a pattern that leverages Enumeration and simulates a rich Enumeration.  @jonathanolson do you think that is good enough to close this issue and change the "Rich enumerations are not currently supported" section of the design patterns doc to refer to it?

**Comment by samreid** [comment]
> Or each Enumeration could be a type, then create values as instances of the type, then instanceof should work...

We have decided not to follow that approach for our Enumeration implementation.  Also, this issue has veered towards support for RichEnumeration, so I'll rename it as such.

**Comment by samreid** [comment]
I experimented with support for rich enumerations, usage is like so:

```js
  class Planet {
    constructor( order ) {
      this.order = order;
    }

    getText( name ) {
      return name + ' is a person from the ' + this.order + ' planet.';
    }
  }

  const Planets = new Enumeration( {
    MARS: new Planet( 2 ),
    EARTH: new Planet( 3 )
  } );

  // Planets.MARS = 'hello'; // fails
  console.log( Planets.MARS.order );
  console.log( Planets.MARS.getText( 'Myself' ) );
```

The implementation overloads the Enumeration constructor like so:

```diff
Index: js/Enumeration.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/Enumeration.js	(revision 72150062806e4919214625020cf800cbf22d242c)
+++ js/Enumeration.js	(date 1574700728707)
@@ -69,6 +69,11 @@
      * @param {Object} [options]
      */
     constructor( keys, options ) {
+      let richSupport = null;
+      if ( !Array.isArray( keys ) ) {
+        richSupport = keys;
+        keys = _.keys( keys );
+      }
 
       assert && assert( typeof options !== 'function', 'options should not be a function' );
 
@@ -106,11 +111,20 @@
       this.KEYS = [];
 
       keys.forEach( key => {
-        this[ key ] = {
-          name: key, // PhET-iO public API relies on this mapping, do not change it lightly
-          toString() {return key;}
-        };
-        this.VALUES.push( this[ key ] );
+        let value = null;
+        if ( richSupport ) {
+          value = richSupport[ key ];
+          value.name = key;
+          value.toString = () => key;
+        }
+        else {
+          value = {
+            name: key, // PhET-iO public API relies on this mapping, do not change it lightly
+            toString() {return key;}
+          };
+        }
+        this[ key ] = value;
+        this.VALUES.push( value );
         this.KEYS.push( key );
       } );
 ```

I'd like to use this feature in Circuit Construction Kit, but I don't want to commit this to master without discussing it first.  @pixelzoom or @jonathanolson do you have time to take a look and discuss?

**Comment by samreid** [comment]
If we decide to move forward with this idea, we can improve it by:
* Updating the JSDoc
* Asserting that the rich enumeration values don't bring their own `name` or `toString` values.

**Comment by pixelzoom** [comment]
I'm not a fan of overloading constructors.  And now the implementation is getting squirrelly.  I.e. in this change:

```js
+        if ( richSupport ) {
+          value = richSupport[ key ];
+          value.name = key;
+          value.toString = () => key;
+        }
+        else {
+          value = {
+            name: key, // PhET-iO public API relies on this mapping, do not change it lightly
+            toString() {return key;}
+          };
+        }
```

.... there are now 2 entirely different (but functionally identical!) ways of setting `value.name` and `value.key`.  And the "PhET-iO public API relies..." comment is only associated with the second way, when it actually is true of _both_ ways of setting `name`. 

Might it be better to have 2 different types, `Enumeration` and `RichEnumeration`?

**Comment by samreid** [comment]
I'm reluctant to create a separate type since most of the implementation will be the same, unless we decide that would be a preferable API.  Here is an implementation that addresses the following issues:

* Constructor arguments are not checked based on type, instead the config pattern is used internally
* The differing ways of setting `value.name` and `value.key` have been unified, and the comment applies to the only occurrence

Proposed usage API is `Enumeration.byKeys()` for plain enumerations and `Enumeration.byMap()` for rich enumerations.  I'd love to discuss better names for these but thought this was clear enough for feedback on the strategy.

Here is the proposed implementation:

<details>

```js
// Copyright 2018-2019, University of Colorado Boulder

/**
 * Creates a simple enumeration, with most of the boilerplate.
 *
 * An Enumeration can be created like this:
 *
 *   const CardinalDirection = Enumeration.byKeys( [ 'NORTH', 'SOUTH', 'EAST', 'WEST' ] );
 *
 * OR using rich values like so:
 *
 *   const CardinalDirection = Enumeration.byMap( {NORTH: northObject, SOUTH: southObject, EAST: eastObject, WEST: westObject} );
 *
 * and values are referenced like this:
 *
 *   CardinalDirection.NORTH;
 *   CardinalDirection.SOUTH;
 *   CardinalDirection.EAST;
 *   CardinalDirection.WEST;
 *
 *   CardinalDirection.VALUES;
 *   // returns [ CardinalDirection.NORTH, CardinalDirection.SOUTH, CardinalDirection.EAST, CardinalDirection.WEST ]
 *
 * And support for checking whether any value is a value of the enumeration:
 *
 *   CardinalDirection.includes( CardinalDirection.NORTH ); // true
 *   CardinalDirection.includes( CardinalDirection.SOUTHWEST ); // false
 *   CardinalDirection.includes( 'NORTH' ); // false, values are not strings
 *
 * Conventions for using Enumeration, from https://github.com/phetsims/phet-core/issues/53:
 *
 * (1) Enumerations are named like classes/types. Nothing in the name needs to identify that they are Enumerations.
 *     See the example above: CardinalDirection, not CardinalDirectionEnum or CardinalDirectionEnumeration.
 *
 * (2) Enumeration values are named like constants, using uppercase. See the example above.
 *
 * (3) If an Enumeration is closely related to some class, then make it a static field of that class. If an
 *     Enumeration is specific to a Property, then the Enumeration should likely be owned by the class that
 *     owns that Property.
 *
 * (4) If an Enumeration is not closely related to some class, then put the Enumeration in its own .js file.
 *     Do not combine multiple Enumerations into one file.
 *
 * (5) If a Property takes an Enumeration value, its validation typically looks like this:
 *
 *     const cardinalDirectionProperty = new Property( CardinalDirection.NORTH, {
 *       validValues: CardinalDirection.VALUES
 *     }
 *
 * (6) Values of the Enumeration are considered instances of the Enumeration in documentation. For example, a method
 *     that that takes an Enumeration value as an argument would be documented like this:
 *
 *     // @param {Scene} mode - value from Scene Enumeration
 *     setSceneMode( mode ) {
 *       assert && assert( Scene.includes( mode ) );
 *       //...
 *     }
 *
 * @author Jonathan Olson <jonathan.olson@colorado.edu>
 */
define( require => {
  'use strict';

  // modules
  const merge = require( 'PHET_CORE/merge' );
  const phetCore = require( 'PHET_CORE/phetCore' );

  class Enumeration {

    /**
     * @param {Object} [config] - must provide keys such as {keys:['RED','BLUE]}
     *                          - or map such as {map:{RED: myRedValue, BLUE: myBlueValue}}
     *
     * @private - clients should use Enumeration.byKeys or Enumeration.byMap
     */
    constructor( config ) {
      assert && assert( config, 'config must be provided' );

      const keysProvided = !!config.keys;
      const mapProvided = !!config.map;
      assert && assert( keysProvided !== mapProvided, 'must provide one or the other but not both of keys/map' );

      const keys = config.keys || Object.keys( config.map );
      const map = config.map || {};

      config = merge( {

        // {string|null} Will be appended to the EnumerationIO documentation, if provided
        phetioDocumentation: null,

        // {function|null} If provided, it will be called as beforeFreeze( enumeration ) just before the
        // enumeration is frozen. Since it's not possible to modify the enumeration after
        // it is frozen (e.g. adding convenience functions), and there is no reference to
        // the enumeration object beforehand, this allows defining custom values/methods
        // on the enumeration object itself.
        beforeFreeze: null
      }, config );

      assert && assert( Array.isArray( keys ), 'Values should be an array' );
      assert && assert( _.uniq( keys ).length === keys.length, 'There should be no duplicated values provided' );
      assert && keys.forEach( value => assert( typeof value === 'string', 'Each value should be a string' ) );
      assert && keys.forEach( value => assert( /^[A-Z][A-Z0-9_]*$/g.test( value ),
        'Enumeration values should be uppercase alphanumeric with underscores and begin with a letter' ) );
      assert && assert( !_.includes( keys, 'VALUES' ),
        'This is the name of a built-in provided value, so it cannot be included as an enumeration value' );
      assert && assert( !_.includes( keys, 'KEYS' ),
        'This is the name of a built-in provided value, so it cannot be included as an enumeration value' );
      assert && assert( !_.includes( keys, 'includes' ),
        'This is the name of a built-in provided value, so it cannot be included as an enumeration value' );

      // @public (phet-io) - provides additional documentation for PhET-iO which can be viewed in studio
      // Note this uses the same term as used by PhetioObject, but via a different channel.
      this.phetioDocumentation = config.phetioDocumentation;

      // @public {string[]} (read-only) - the string keys of the enumeration
      this.KEYS = keys;

      // @public {Object[]} (read-only) - the object values of the enumeration
      this.VALUES = [];

      keys.forEach( key => {
        const value = map[ key ] || {};

        // Set attributes of the enumeration value
        assert && assert( value.name === undefined, 'rich enumeration values cannot provide their own name attribute' );
        assert && assert( value.toString === Object.prototype.toString, 'rich enumeration values cannot provide their own toString' );
        value.name = key; // PhET-iO public API relies on this mapping, do not change it lightly
        value.toString = () => key;

        // Assign to the enumeration
        this[ key ] = value;
        this.VALUES.push( value );
      } );

      config.beforeFreeze && config.beforeFreeze( this );
      assert && Object.freeze( this );
      assert && Object.freeze( this.VALUES );
      assert && Object.freeze( this.KEYS );
    }

    /**
     * Checks whether the given value is a value of this enumeration. Should generally be used for assertions
     * @public
     *
     * @param {Object} value
     * @returns {boolean}
     */
    includes( value ) {
      return _.includes( this.VALUES, value );
    }

    /**
     * Creates an enumeration based on the provided string array
     * @param {string[]} keys - such as ['RED','BLUE']
     * @param {Object} [options]
     * @returns {Enumeration}
     * @public
     */
    static byKeys( keys, options ) {
      assert && assert( !options || options.keys === undefined );
      return new Enumeration( merge( { keys: keys }, options ) );
    }

    /**
     * Creates a "rich" enumeration based on the provided map
     * @param {Object} map - such as {RED: myRedValue, BLUE: myBlueValue}
     * @param {Object} [options]
     * @returns {Enumeration}
     * @public
     */
    static byMap( map, options ) {
      assert && assert( !options || options.map === undefined );
      return new Enumeration( merge( { map: map }, options ) );
    }
  }

  return phetCore.register( 'Enumeration', Enumeration );
} );

```

</details>

And here is a demonstrative unit test:

<details>

```js
  QUnit.test( 'Rich', function( assert ) {
    class Planet {
      constructor( order ) {
        this.order = order;
      }

      getText( name ) {
        return name + ' is a person from the ' + this.order + ' planet.';
      }
    }

    const Planets = Enumeration.byMap( {
      MARS: new Planet( 2 ),
      EARTH: new Planet( 3 )
    } );

    assert.ok( Planets.MARS.order === 2, 'mars order should match' );
    assert.ok( typeof Planets.EARTH.getText( 'bob' ) === 'string', 'should return a string' );
    window.assert && assert.throws( () => {
      Planets.MARS = 'hello'; // fails because enumeration values should not be reassignable
    } );
  } );
```

</details>

Once we move forward with this (or something like it), we will be able to replace the `beforeFreeze` pattern with the rich enumeration pattern, and then delete `beforeFreeze` altogether. 

@pixelzoom can you please take a look?

**Comment by pixelzoom** [comment]
I have not reviewed the previous comment. I'd like to hear @jonathanolson's thoughts before spending more time on this issue.

**Comment by jonathanolson** [comment]
The above looks good to me (although we may want to ensure that rich values don't have a `name` already existing that could get overwritten). Additionally, this now does NOT freeze the rich enumeration objects (or any actual enumeration values). Is it important to be able to freeze things, so `SomeMode.SOMETHING.name = 'foobar'` would be prevented after the enumeration is constructed?

That's also an issue with the current enumeration pattern, so I think it's good to move forward with your changes.

**Comment by samreid** [comment]
> (although we may want to ensure that rich values don't have a name already existing that could get overwritten)

That is checked on this line of the proposal

```js
assert && assert( value.name === undefined, 'rich enumeration values cannot provide their own name attribute' );
```

> Additionally, this now does NOT freeze the rich enumeration objects (or any actual enumeration values). Is it important to be able to freeze things, so SomeMode.SOMETHING.name = 'foobar' would be prevented after the enumeration is constructed?

I believe that can be accomplished with this line:
```js
assert && keys.forEach( key => assert && Object.freeze( map[ key ] ) );
```

And I tested that with this new part of the unit test:

```js
    window.assert && assert.throws( () => {
      Planets.MARS.name = 'not mars!'; // Should not be able to reassign enumeration value properties
    } );
```

**Comment by pixelzoom** [comment]
Proposed implementation in https://github.com/phetsims/phet-core/issues/50#issuecomment-561494563 looks good to me. 

Is there (or should there be) verification that all values of a rich enumeration are of the same type?

**Comment by samreid** [comment]
> Is there (or should there be) verification that all values of a rich enumeration are of the same type?

There is not currently a verification for that.  However, I'm wondering if we should support use cases like the following, where the `constructor` values are different:

```js
Enumeration.byMap({
   mercury: new Planet(),
   earth: new Earth() // Earth extends Planet
});
```

If we want to support that, then all we need to check is that all of the objects share a common ancestor.  Which I think will generally be true (if everything extends Object).

**Comment by samreid** [comment]
Thanks, I committed the implementation with revisions noted above.  Leaving assigned to @pixelzoom to finish discussing https://github.com/phetsims/phet-core/issues/50#issuecomment-562359527 and I'll open a new issue for investigating whether we can now eliminate `beforeFreeze`.

**Comment by pixelzoom** [comment]
Re https://github.com/phetsims/phet-core/issues/50#issuecomment-562359527...  I've never encountered heterogeneous enumeration values.  What's the use case?  What do other developers think?   The point of enumerations is that their values are supposed to be opaque.  Heterogenous values means different APIs for different values, and that seems odd (and maybe indicative of a case where you shouldn't be using an enum.)

**Comment by jonathanolson** [comment]
It seems cleaner to put data on enumeration values than creating functions to compute/extract values.

The most obvious alternative then is:
```js
const coordinateMap = {
  [ Orientation.HORIZONTAL ]: 'x',
  [ Orientation.VERTICAL ]: 'y',
};
Orientation.getCoordinate( orientation ) { return coordinateMap[ orientation ]; }
// callsite
Orientation.getCoordinate( orientation );
```
as opposed to:
```js
this.coordinate = // the constructor parameter
// callsite
orientation.coordinate;
```

**Comment by zepumph** [comment]
If I understand correctly, I think I tend to agree more with @jonathanolson and @pixelzoom than with @samreid. I can't think of a use case in which this would be needed, and couldn't be solved by using `beforeFreeze`. That said maybe I'm to ingrained in the current pattern, but I see nothing wrong with `PhotonTarget`, and don't need to see a data structure encapsulate the molecule name inside each enum value.

**Comment by jbphet** [comment]
The question of whether the enumeration values all need to be the same type was discussed in the 1/1/2020 developer meeting.  The current thinking is that yes, they should be the same type.  However, there may be a point in the future where it would make sense to have them share a common base class but not be exactly the same class.  If that arises, we will revisit, and may do something like provide an optional "sharedBaseClass" value against which all would be validated.

During the same dev meeting we reviewed the way that the enum can end up using string access of object properties.  Examples can be seen in `PartitionSizeEditNode`.  We concluded that this usage is acceptable.

Assigning to @samreid for followup.
